Convert tactics chapter to prodn, part 2 (#16498)
coq/coq: doc/sphinx/proofs/writing-proofs/proof-mode.rst

for which the type of their type is `Prop` or `SProp`,
but these terms are also used interchangeably.

> .. coqtop:: all

Reviewer: Is there a good reason to change this from `out` to `all`? It seemed to me that the implementation details were not made to be shown here...

Author: It seemed mysterious to hide the small trick needed to print the type.  I could do a parenthetical comment instead.  OTOH, how often do users need to print the type?

Reviewer: > OTOH, how often do users need to print the type?

Not often. Only users writing advanced automation would do that. So the example was just meant to highlight the type of the hypotheses not how we compute that.

   Show Existentials.

Users can control which goals are displayed in the context by :term:`focusing <focus>` and

> _shelving_ goals.  Focusing lets the user pick a single goal to work on.  Focusing operations

Reviewer: You probably meant to use italic here. That would be with `*` instead of `_`.

Reviewer: They are very often not useful. I call on your personal judgment to decide when to keep them or not.

can be nested.  See :ref:`focused_goals`.  Goals can be :term:`shelved` so they're not shown

> in the context until they are later :term:`unshelved`.  See :ref:`shelved_goals`.

Reviewer: "so they're not shown in the context" seems to indicate that this is the purpose of shelving. The main reason for shelving goals is when they are expected to be solved by side effects (unification of existential variables happening while running tactics). Marginally, shelving can be used as a trick for other reasons.

> 

Author: Q: All goals are existential variables.  Is it also true that all existential variables are goals?  I'd like to point this out or contrast them.  And IIUC different from `(exists x ...` variables.

Reviewer: > Is it also true that all existential variables are goals?

Yes, it is true. Most of the time existential variables correspond to shelved goals. On the other hand, we often use the term goals to designate the goals that we work on directly, whereas shelved goals are most often resolved by a side effect.

> And IIUC different from `(exists x ...` variables.

Right. Note that a connection exists though. If you have a goal of the form `exists x, A`, then the tactic `eexists` generates a new existential variable corresponding to `x` (say `?x`) and leaves the subgoal `A`.

Author: > shelved goals are most often resolved by a side effect

I'm not familiar with that, that's worth describing (e.g. a paragraph + example that applies to all the e* versions of tactics).  Maybe you can provide a realistic example?  The following is probably unrealistic if not entirely on the wrong track.  I don't know if/how the proof can be completed:

```coq
Goal exists n : nat, n = n.
eexists ?[x].
  (*
  1 subgoal
  ______________________________________(1/1)
  ?x = ?x
  *)
Show Existentials.
  (*
  Existential 1 = ?x : [ |- nat] (shelved)
  Existential 2 = ?Goal : [ |- ?x = ?x]
  *)
reflexivity.
  (*
  All the remaining goals are on the shelf:

  nat
  *)
```

Reviewer 2: This one doesn't solve the `n : nat` goal since you still need to provide a witness. What would be "solving existential by side effect" is if you had this:
```coq
Goal forall (a : nat), a = a -> exists n : nat, n = n.
Proof.
  intros a e.
  eexists ?[x].
  exact e.
Qed.
```

Now you can see that the existential variable (subgoal) n : nat can be solved by inferring from the data in `e : a = a`.

Reviewer: >shelved goals are most often resolved by a side effect

Another way of explaining this is: a rule of thumb is that tactics create shelve goals by default when they correspond to existential variables that appear somewhere in one of the (non-shelved) subgoals. Similar to what the `shelve_unifiable` tactic would do. This is so because usually proving these goals will instantiate these existential variables in passing (what I call a side effect).

Author: This seems like a bug ("clear e" shouldn't remove e because the goal relies on it).  Right?

```coq
Goal forall (a : nat), a = a -> exists n : nat, n = n.
Proof.
intros a e.
eexists ?[x].
  (* 
    1 subgoal
    a : nat
    e : a = a
    ______________________________________(1/1)
    ?x@{a:=a; e:=e} = ?x@{a:=a; e:=e}
  *)
clear e.
  (*
    1 subgoal
    a : nat
    ______________________________________(1/1)
    ?x@{a:=a} = ?x@{a:=a}
  *)
```

How often do users need to explicitly instantiate existential variables rather than rely on side effects?

Author: Is there a tactic that will prove `Goal nat.`?

Reviewer: >I had "Display existential variable instances" enabled in CoqIDE.

That's not something you want to activate in general. This will just print the context of the existential variable, so everything that it *could* eventually depend on when instantiated. However, this does not make the goal depend on all of this context and thus the behavior of `clear` is perfectly OK here.

>How often do users need to explicitly instantiate existential variables rather than rely on side effects?

Rarely.

> Is there a tactic that will prove `Goal nat.`?
> 
> I guess that's just `apply O` or `apply (S (O))` ...

Yep. Or just `exact 0`.

   Show Existentials.

Users can control which goals are displayed in the context by :term:`focusing <focus>`
goals.  Focusing lets the user (initially) pick a single goal to work on.  Focusing

> operations can be nested.  Tactics creating existential variables create

Author: How does this look?

```
Users can control which goals are displayed in the context by :term:`focusing <focus>`
goals.  Focusing lets the user (initially) pick a single goal to work on.  Focusing
operations can be nested.

Tactics such as :tacn:`eapply` create existential variables as placeholders for
undetermined variables as well as creating :term:`shelved <shelved>` goals.
Shelved goals are not shown in the context by default, but they can be unshelved
to make them visible.  Other tactics may automatically resolve these goals
(whether shelved or not); the purpose of shelving is to hide goals that the
user usually doesn't need to think about.  See :ref:`existential-variables`
and :ref:`this example <automatic-evar-resolution>`.
```

> operations can be nested.  Tactics creating existential variables create

Reviewer: Perfect!

   .. seealso:: :cmd:`Proof with`

   .. todo PR: Add a description for Proof Mode?

.. cmd:: Proof Mode @string

>    :undocumented:

Author: Yes, a bug.  BTW, you do realize that Ltac2 doesn't support `idtac`?  Other thoughts:
- Not sure changing `Proof Mode` in the middle of a proof is a good idea or that it's a great command to have
- Probably should have been `Set Proof Mode`

```
From Ltac2 Require Import Ltac2.

Goal True.
Fail idtac. (* Unbound value idtac *)
```

>    :undocumented:

Reviewer: > BTW, you do realize that Ltac2 doesn't support `idtac`?

That's on purpose. The Ltac2 doc says:

>idtac is no more. Either use () if you expect nothing to happen, (fun () => ()) if you want a thunk (see next section), or use printing primitives from the Message module if you want to display something.

>Not sure changing `Proof Mode` in the middle of a proof is a good idea or that it's a great command to have

It's probably not very much used (since it is not documented), but I think this is not completely absurd to do that.

>Probably should have been `Set Proof Mode`

But then it would not respect the default locality behavior of options. There are no options whose scope is limited to the current proof (although maybe they should).

>AFAIK there's no way to show the current proof mode

That's unfortunate, but we have `Test Default Proof Mode` to know the default proof mode and if you are in the middle of a proof, you can look at the proof to see if there are any uses of the `Proof Mode` command.

   .. exn:: No focused proof to restart.
      :undocumented:

.. _focused_goals:

Focusing goals
``````````````

:gdef:`Focusing <focus>` lets you limit the context display to (initially) a
single goal.  If a tactic creates additional goals from a focused goal, the
subgoals are also focused.  The two focusing constructs are
:ref:`curly braces <curly-braces>` (`{` and `}`) and :ref:`bullets <bullets>`
(e.g. `-`, `+` or `*`).  These constructs can be nested.

.. _curly-braces:

> Curly braces

Reviewer: Curly braces and curly brackets are equivalent, right? I see that we are not completely consistent in the terminology. At least one error message and its description below use "brackets". Not a big deal though.

Author: The Internet says "The term "curly braces" is more favored in the U.S., while "brackets" is more widely used in British English.". I had never seen them referred to as "curly brackets".

Reviewer: OK, let's try to consistently use "braces" everywhere in the docs and we can postpone fixing the error messages.

Reviewer: So in US English it's curly braces and square brackets?

Author: In my experience, yes, but that may not be universal.  The Unicode table, which is probably meant to be American English, describes "{" as "left curly bracket".  The wikipedia article has the following chart, which has merely "braces".

![image](https://user-images.githubusercontent.com/1253341/193866849-f7fd9360-9e3e-42e3-b448-7127e71c1b23.png)

It also says "In most English-speaking countries, an unqualified word "bracket" refers to the parenthesis (round bracket); in the United States, the square bracket."

So I guess it should be "braces" or "curly brackets".

Reviewer 2: I personally have never heard people use "curly bracket" in the UK.

   .. exn:: No focused proof to restart.
      :undocumented:

.. _focused_goals:

> Focusing goals

Author: I made them "Curly brackets" and "Bullets" under "Focusing Goals".  They are level 5 titles, so they don't appear in the contents panel.

![image](https://user-images.githubusercontent.com/1253341/193715415-76580742-187e-4200-9ad4-8a3d50b1c3ce.png)

   :n:`@natural:`
     Focuses on the :token:`natural`\-th goal to prove.

.. _focus_shelved_goal:

   :n:`[ @ident ]: %{`
     Focuses on the goal named :token:`ident` even if the goal is not in focus.

>      Goals are :ref:`existential-variables`, which don't have names by default.

Reviewer: ```suggestion
     Goals are :term:`existential variables <existential variable>`, which don't have names by default.
```

Author: Hmm, that suggests you'd favor dropping the "See X" I mention in https://github.com/coq/coq/pull/16498#discussion_r983886282

Reviewer: Yes, that's a typical example where I don't think linking to the section is useful.

     You can give a name to a goal by using :n:`refine ?[@ident]`.

   .. example:: Working with named goals

      .. coqtop:: in

         Ltac name_goal name := refine ?[name].  (* for convenience *)
         Set Printing Goal Names.  (* show goal names, e.g. "(?base)" and "(?step)" *)

      .. coqtop:: all

         Goal forall n, n + 0 = n.
         Proof.
         induction n; [ name_goal base | name_goal step ].
         (* focus on the goal named "base" *)
         [base]: {
           reflexivity.

      .. coqtop:: in

         }

>       .. todo PR: Following seems unrealistic, maybe should be a separate example?

Reviewer: This simplified case is unrealistic, but it can happen that existential variables remain unsolved when all goals are solved and this is the most realistic example I could find that is also sufficiently simple.

   indentation. For example ``-`` must be the outer bullet and ``**`` the inner

>    one in the example below.

Reviewer: The note does not match anymore the modified example.

Author: Got it.

FWIW, your comment reads oddly (ungrammatically?) because "anymore" is out of place.  It should be "The note does not match the modified example **anymore**."  I see similar mistakes by others.  Does French expect adverbs to be next to the verb they modify?

Reviewer: Yes, I think that's very likely a Frenchism. In this case, the French version would be "ne correspond plus" (standard negative construction).

nesting levels provided they are delimited by curly braces. Bullets are made from
``-``, ``+`` or ``*`` characters (with no spaces and no period afterward):

.. tacn:: @bullet

>    :undocumented:

Author: In this case it's needed. Some Sphinx bug.

   - "None": this makes bullets inactive.
   - "Strict Subproofs": this makes bullets active (this is the default behavior).

> .. cmd:: Unfocused

Reviewer: Can you add an "Other focusing commands" or whatever section so that they do not end up inside the "Set Bullet Behavior" section?

   You tried to apply a tactic but no goals were under focus.
   Using :n:`@bullet` is  mandatory here.

.. exn:: No such goal. Try unfocusing with %}.

   You just finished a goal focused by ``{``, you must unfocus it with ``}``.

> .. note:: Use :opt:`Default Goal Selector` with the ``!`` selector forces

Reviewer: ```suggestion
.. note:: Using :opt:`Default Goal Selector` with the ``!`` selector forces
```

   tactic scripts to keep exactly one goal focused (e.g. using bullets)

>    or use explicit goal selectors.

Reviewer: ```suggestion
   or to use explicit goal selectors.
```

Author: No, should not be the infinitive.  Here's what I put (with "to force"):

```
.. note:: Use :opt:`Default Goal Selector` with the ``!`` selector to force
   tactic scripts to keep exactly one goal focused (e.g. using bullets)
   or use explicit goal selectors.
```

Reviewer: But is it clear then that the parentheses are this way:

>Use :opt:`Default Goal Selector` with the ``!`` selector to [force tactic scripts to keep exactly one goal focused (e.g. using bullets) or use explicit goal selectors].

and not this way:

>[Use :opt:`Default Goal Selector` with the ``!`` selector to force tactic scripts to keep exactly one goal focused (e.g. using bullets)] or [use explicit goal selectors].

I find it ambiguous when reading this.

Author: It's not very clear.  How about something more like:

Use `Set Default Goal Selector !` to ensure that each tactic is applied to a single goal at a time.  This will cause a failure if more than one goal is in focus.  Less conveniently, you can also add goal selectors such as `1 :` where needed.

OTOH, the default goal selector is `1` (the initial setting of Default Goal Selector), which also ensures that the tactic is only applied to a single goal at a time (unless you have a selector). So maybe this is just a useless note that should be dropped?

Reviewer: No, your new wording is less good I think. Maybe you have misunderstood the point of `!` (at least this is what I gather from your question).

The point of `Set Default Goal Selector !` is not to ensure that each tactic is applied to a single goal, not at all: it doesn't even prevent applying tactics to several goals (with goal selectors).

`Set Default Goal Selector !` forces the use of focusing mechanisms (bullets, braces) and goal selectors so that it is always explicit to which goal(s) a tactic is applied.

Does this help?

   Performs :n:`@tactic`, then unshelves existential variables added to the
   shelf by the execution of :n:`@tactic`, prepending them to the current goal.

.. tacn:: admit
          give_up

   Allows skipping a subgoal to permit further progress on the rest of the
   proof.  The selected goals are removed from the context.  They are not
   solved and cannot be solved later in the proof. Since the goals are not
   solved, the proof cannot be closed with :cmd:`Qed` but only with :cmd:`Admitted`.

>    :n:`give_up` is a synonym for :n:`admit`.

Reviewer: I think this sentence does not need to be kept.
```suggestion
```
